### ISSUE TITLE
Receiver Close() times out due to block in receive

### ISSUE THREAD

**Issue by princjef** [Azure/go-amqp#38]
I am using this package through the https://github.com/Azure/azure-event-hubs-go SDK to continuously receive messages from an event hub.

When the system experiences extreme load and/or occasional network instability, calls to `Close()` on an individual Event Hubs receiver (and by extension, its underlying receiver/link from this package) have a tendency to hang until the provided context expires. My typical context timeout is 10 seconds, but I've observed the same with much longer timeouts (it doesn't appear to be related to time).

I recently captured a goroutine dump of a process that was encountering this problem and observed that each timed out `Close()` was causing an extra goroutine to be left dangling indefinitely (along with the full prefetch buffer for the associated link). These are the relevant groups of leaky goroutines:

```
1522 @ 0x43ab25 0x4059fa 0x405755 0xa70197 0xa70df8 0xa6f47f 0x4717c1
#	0xa70196	github.com/Azure/go-amqp.(*link).muxReceive+0x276	/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:513
#	0xa70df7	github.com/Azure/go-amqp.(*link).muxHandleFrame+0x117	/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:548
#	0xa6f47e	github.com/Azure/go-amqp.(*link).mux+0x37e		/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:304

386 @ 0x43ab25 0x4059fa 0x405755 0xa70197 0xa70df8 0xa6fe35 0xa6f2d6 0x4717c1
#	0xa70196	github.com/Azure/go-amqp.(*link).muxReceive+0x276	/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:513
#	0xa70df7	github.com/Azure/go-amqp.(*link).muxHandleFrame+0x117	/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:548
#	0xa6fe34	github.com/Azure/go-amqp.(*link).muxFlow+0x234		/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:380
#	0xa6f2d5	github.com/Azure/go-amqp.(*link).mux+0x1d5		/go/pkg/mod/github.com/!azure/go-amqp@v0.13.4/link.go:290
```

When inspecting the source for the version I'm using (0.13.4), the line where those goroutines are hung is the last line of this block:

https://github.com/Azure/go-amqp/blob/2957005aeaa725796bbd4ae84de0dd5bf58d5fa7/link.go#L510-L513

As the comment mentions, the code is designed never to block here (due to the channel being buffered for the exact number of messages in the link credit), but something is clearly violating that assumption.

I'm not sure whether the issue is a protocol violation by Event Hubs w.r.t. link credit or a bug causing extra messages to be buffered in the channel. Either way, it seems dangerous to assume that everything will be perfectly well-behaved all the time. It seems like a simple `select` to handle the shutdown case would solve the problem here. Happy to send a PR if it seems like a reasonable change.

**Comment by jhendrixMSFT** [comment]
It would be interesting if you could get a snapshot of goroutines when `Close()` hangs (I assume there were others than what was pasted).  Based on the provided info, I believe the hang is coming from the call to [`c.getErr()`](https://github.com/Azure/go-amqp/blob/master/conn.go#L290) while closing the connection.  This will block until [`conn.mux()`](https://github.com/Azure/go-amqp/blob/master/conn.go#L334) exits which _should_ be in progress now that `c.closeMux` has been closed.

**Comment by jhendrixMSFT** [comment]
Also, if you have a repro that would be reasonable for me to run I'm happy to take a look.

**Comment by jhendrixMSFT** [comment]
Fixed in post-v0.17.0
